torch/csrc/jit/serialization.h

```diff
@@ -0,0 +1,285 @@
+     throw std::runtime_error("Magic number mismatch in PyTorch file. File may"
+                              " be corrupted or is not actually a PyTorch file.");
+   }
+   uint64_t file_format_version = read64BitIntegerLittleEndian();
+   if (file_format_version > kMaxSupportedFileFormatVersion) {
+     std::ostringstream errmsg;
+     errmsg << "Attempted to read a PyTorch file with version " << file_format_version
+            << " but the maximum supported version for reading is " << kMaxSupportedFileFormatVersion
+            << ". Your PyTorch installation may be too old.";
+     throw std::runtime_error(errmsg.str());
+   }
+   seekToNextAlignmentBoundary();
+  }
+  void readAndValidateFileFooter() {
+    // Seek to location of file footer. We've already validated that the file
+    // length is a multiple of the alignment size
+    cursor = file_size - kFieldAlignment;
+    std::fseek(fp, cursor, SEEK_SET);
+    auto tag = read64BitIntegerLittleEndian();
+    if (tag != RecordTags::FOOTER) {
+      throw std::runtime_error("File footer has wrong record type. Is this"
+                               " file corrupted?");
+    }
+    last_record_offset = read64BitIntegerLittleEndian();
+    if (last_record_offset > file_size) {
+      throw std::runtime_error("Offset of last record is higher than the size"
+                               " of the file! Is this file corrupted?");
+    }
+  }
+};
+
+class PyTorchFileWriter {
+ public:
+  PyTorchFileWriter(const std::string& filename) {
+    fp = std::fopen(filename.c_str(), "wb");
+    if (!fp) {
+      wrapPErrorAndThrow("Unable to open PyTorch file for writing!");
+    }
+    writeFileHeader();
+    // In the case that we do not write any records into this file, the last
+    // record index written into the footer will point to the footer itself.
+    last_record_idx = cursor;
+  }
+  uint64_t writeRecord(const char* data, size_t size) {
+    JIT_ASSERT(!finalized);
+    uint64_t record_offset = cursor;
+    last_record_idx = record_offset;
+    write64BitIntegerLittleEndian(RecordTags::STORAGE);
+    write64BitIntegerLittleEndian(size);
+    padToNextAlignmentBoundary();
+    writeBuffer(data, size);
+    padToNextAlignmentBoundary();
+    return record_offset;
+  }
+  void writeEndOfFile() {
+    JIT_ASSERT(!finalized);
+    writeFileFooter();
+    finalized = true;
+    std::fclose(fp);
+  }
+  ~PyTorchFileWriter() {
+    if (!finalized) {
+      writeEndOfFile();
+    }
+  }
+ private:
+  FILE *fp;
+  size_t cursor = 0;
+  bool finalized = false;
+  size_t last_record_idx = 0;
+
+  // Utility functions
+  void write64BitIntegerLittleEndian(const uint64_t value) {
+    // TODO endian swap on platforms that need it?
+    if (!std::fwrite(&value, 8u, 1u, fp)) {
+      wrapPErrorAndThrow("Unable to write to file!");
+    }
+    cursor += 8u;
+  }
+
+  void writePad(const size_t num_bytes) {
+    uint8_t pad_val = kPadValue;
+    // TODO is there a more efficient way to do this?
```

Comment:
```
static std::vector<char> zeros('\0', 64);
std::fwrite(&zeros.begin()...
```

Reply:
butbutbut i want the kernel to do that for me :'(